nmtwizard/preprocess/prepoperator.py

```diff
@@ -22,105 +22,147 @@ def get_operator_params(config):
+
+        # TODO : do we really need those now ?
+        self.start_state = { "src_tok_config" : None,
+                             "tgt_tok_config" : None }
+
+        # Current state of pipeline.
+        # Passed to and modified by operator initializers if necessary.
+        self._build_state = dict(self.start_state)
+
+        self._build_preprocess_pipeline(config)
+
+
+
+class PostprocessPipeline(InferencePipeline):
+
+    def __init__(self, config):
+
+        super(PostprocessPipeline, self).__init__(config, ProcessType.POSTPROCESS)
+
+        # Reverse preprocessing operators.
+        self._ops = reversed(self._ops)
+
+        # TODO
+        self.start_state, self._build_state = self._build_state, self.start_state
+        # Add pure postprocessing operators.
+        postprocess_config = config.get("postprocess")
+        if postprocess_config:
+            for op in postprocess_config:
+                operator = self._build_operator(op)
+                if operator:
+                    self._ops.append(operator)
+
+
 @six.add_metaclass(abc.ABCMeta)
 class Operator(object):
     """Base class for preprocessing operators."""
 
+    def __call__(self, tu_batch, process_type):
+        if self._is_enabled_for(process_type):
+            tu_batch = self._apply_operator(tu_batch, process_type)
+        return tu_batch
+
     @abc.abstractmethod
-    def __call__(self, tu_batch):
+    def _apply_operator(self, tu_batch, process_type):
         raise NotImplementedError()
 
-    def is_applied(self):
-        # Operator is applied by default, unless _is_applied property is explicitely set to False.
-        if hasattr(self, "_is_applied"):
-            return self._is_applied
+    def _is_enabled_for(self, process_type):
         return True
 
 
 class TUOperator(Operator):
     """Base class for operations iterating on each TU in a batch."""
 
-    def __call__(self, tu_batch):
-
+    def _apply_operator(self, tu_batch, process_type):
         # TU operator applies an action to each tu.
         # The action yields zero, one or more element for the new list
-        tu_batch = list(chain.from_iterable(self.apply(tu) for tu in tu_batch))
+        tu_batch = list(chain.from_iterable(self._apply_tu_operator(tu, process_type) for tu in tu_batch))
 
         return tu_batch
 
     @abc.abstractmethod
-    def apply(self, tu_batch):
+    def _apply_tu_operator(self, tu, process_type):
         raise NotImplementedError()
 
 
 class Filter(TUOperator):
 
-    def __init__(self, process_type):
+    def __init__(self):
         # TODO: Sub-criteria for source_detok, target_detok, source_tok, target_tok, or both with alignment ?
         self._criteria = []
-        if process_type=="inference" or process_type=="postprocess":
-            self._is_applied = False
 
-    def apply(self, tu):
+    # TODO : make generic ?
+    def _is_enabled_for(self, process_type):
+        if process_type > ProcessType.TRAINING:
```

Comment:
I'm not sure we want to assume an order here. Just return `process_type == ProcessType.TRAINING`.